A container can join pid namespace of another container.

dqminh opened the pull request:
Fix #604 

This supports setting `configs.NEWPID` to an existing PID namespace for a container's init process. It leverages the same C code infrastructure for `execin` to clone a new init process with the wanted PID namespace.

Edit: Because of some sharing in code between init and execin, execin also supports joining userns of the init process now. Init process cant join mnt/userns of another init process yet.

dqminh committed:
set custom namespaces for init process

An init process can join other namespaces (pidns, ipc etc.). This leverages
C code defined in nsenter package to spawn a process with correct namespaces
and clone if necessary.  When a shared-pidns container exits,
the original container will keep running.

This removes joinExistingNamespaces in Go layer because all namespaces will be
setns in C layer.

This also changes setns process to requires a list of namespaces to be joined
rather than requires only the pid of init process and deduct the namespace from
it.

Signed-off-by: Daniel, Dao Quang Minh <dqminh89@gmail.com>

dqminh committed:
use custom id as cgroups name for nsinit

if nsinit specifies `--id` option, use that id as cgroups' name instead of
the directory name. It's helpful where we want to start multiple containers
from the same directory for example.

Signed-off-by: Daniel, Dao Quang Minh <dqminh89@gmail.com>

mrunalp commented on nsenter/nsexec.c:
```diff
@@ -60,6 +60,62 @@ static int clone_parent(jmp_buf * env)
 	return child;
 }
 
+void nsinit_pid()
```
Could we combine this logic into nsexec() itself?

dqminh replied:
i tried that but the end result wasnt as clean. Since spawning an init process with a known pid namespace and exec a new process inside another container are two very different operations , i think that having two functions describes the intent better.

avagin commented:
The same hack is required for mount and user namespaces, because a multithreaded process can't change them.
http://man7.org/linux/man-pages/man2/setns.2.html
'''
A process may not be reassociated with a new mount namespace if it is multithreaded.
'''

What do you think about creating a directory with namespaces in container's root directory. We can bind-mount all required namespaces there before starting the container and then bind-mount container's namespace files to use them for executing processes inside CT. In this case we can use the same code of nsexec.c for starting the first and other processes.

LK4D4 commented:
I agree with @avagin that bind-mounting namespace can be right way to do such stuff, because `nsexec.c` supposed to be very small layer only for setting namespaces and expanding it can lead us to support pretty big C code in golang runtime, which probably will hit some limitations.
I'm not against merging this though if this feature is needed by someone right now :)

dqminh commented:
@LK4D4 There's a few related issues in docker https://github.com/docker/docker/pull/13453 and https://github.com/docker/docker/issues/10163. Overall i think it's better to do this properly (if this is not the best way) since 1.7 is already in RC now so we have time :)

@avagin I dont understand the part about bind-mounting namespaces. Do you have an example for that ? My current understanding of that is that we have a special directory that contains namespace descriptors that we can detect when starting the container and use them to setns with some code from nsexec.c ?

mrunalp commented:
@dqminh The suggestion is to bind mount /proc/<pid>/ns/* files into some directory for each container (Just like how ip netns add works. It bind mounts net namespace fds under /var/run/netns). After that is done, new processes could use those paths to join the container. This could also be used for joining subset of other container's namespaces.

@avagin could correct me if I am wrong.

mrunalp commented:
@dqminh Example https://gist.github.com/mrunalp/5290c92b32fe709aadd3

LK4D4 commented:
Also, you can destroy original namespace, so bind-mounting is like "copy" namespace.

mrunalp commented:
Right, as long the mounted namespace fds exists, the original container can just die without affecting those that joined those namespace fds.

mrunalp commented:
Not sure how that will behave will pid namespaces though since pid 1 going down brings down the other processes.

dqminh commented:
> The suggestion is to bind mount /proc/self/ns/* files into some directory for each container (Just like how ip netns add works

Ah i see. But that would not change the current implementation right, since all libcontainer needs to know is still the paths to the namespace descriptors.

> Not sure how that will behave will pid namespaces though since pid 1 going down brings down the other processes.

i think pid 1 going down will kill the rest of the processes, and you cant join using the open file descriptor anymore (http://man7.org/linux/man-pages/man7/pid_namespaces.7.html).

avagin commented:
> Ah i see. But that would not change the current implementation right, since all libcontainer needs to know is still the paths to the namespace descriptors.
@dqminh I don't understand what you mean. Could you elaborate?

LK4D4 commented:
@dqminh for main process we use Cloneflags for creating namespaces. New idea is to create namespaces separately and then pass them even for first process.

avagin commented:
@LK4D4 @dqminh I don't suggest to create namespaces separately. For example, It's impossible for pidns. For the init process we will mount all existent namespaces there and set clone flags for others.

LK4D4 commented:
@avagin What point of bind-mounting then? Just to collect them in one directory?
I think the we can pass just space-separated namespace list to nsexec.c where it will be generated from `config.Namespaces` for init and from `state.NamespacePaths` for other.

dqminh commented:
Right, we cant really seperate the creation of namespaces and processes at least for pid namespace. It always requires an anchor process.

The current way joining namespaces works for init process is that container has `config.Namespaces` which keys values mapping of namespace to their descriptor path. The value can either be empty ( which then the key is used in CloneFlags ), or points to a valid namespace descriptor ( which we use to join namespace in `joinExistingNamespaces` ). So from libcontainer, the only requirements for init to join different namespaces is to set the key-value pair in `config.Namespaces`

This works for net ( the only current usecase afaik ), but doesnt work for pid ( and probably mount/userns as @avagin pointed out ). This patch adds another intermediate clone to put the init process in correct pid namespaces.

My understanding is that even if we bind-mount namespaces, we will still pass the mount destination as value to `config.Namespaces` and perform `setns` before exec-ing the target process, so it will not be different from the current way.

> I think then we can pass just space-separated namespace list to nsexec.c where it will be generated from config.Namespaces for init and from state.NamespacePaths for other.

Hmm yes, that is another way to do it so we can reuse that blocks of code between init and execin. It also helps when we want to join another namespaces besides pid. Probably joinExistingNamespaces can be removed in that case.

avagin commented:
> @avagin What point of bind-mounting then? Just to collect them in one
directory?

* collect in one directory
* be independent from the init process. If we have a container without
pidns, we will able to enter into it even when the first process died. For
container with pidns, this approach avoids a problem when the init process
died and another process is created with the same pid.

I'm not sure that we need to do this. It's just an idea how it can be fixed.

> I think the we can pass just space-separated namespace list to nsexec.c
where it will be generated from config.Namespaces for init and from
state.NamespacePaths for other.

Yes, we can.

dqminh commented:
@LK4D4 @avagin i added some code to pass a list of namespaces to be joined to both init process and execin process. So they behave the same now on `setns` part.

All `setns` now is done in C layer (`joinExistingNamespaces` is not necessary)

avagin commented on nsenter/nsexec.c:
```diff
@@ -103,46 +95,37 @@ void nsexec()
-		fd = openat(tfd, namespaces[i], O_RDONLY);
+	char *ns, *saveptr;
+	while ((ns = strtok_r(nspaths, ",", &saveptr))) {
+		fd = open(ns, O_RDONLY);
```
If you enter into a mount namespace, you can't be sure that you will able to enter in other namespaces, because nspaths may be inaccessible from this mount namespace.

dqminh replied:
right, I think the order of nspaths has to be set from the caller ( which we are not doing now for init, but we do for exec ). wdyt ?

avagin replied:
I agree that a caller can do this.

dqminh replied:
done. The order of namespaces is now enforced in `orderNamespacePaths`. I also tried to add NEWUSER there but some tests failed. Added a note on NEWUSER and will try again in another patch only for NEWUSER.

avagin commented on container_linux.go:
```diff
@@ -169,21 +168,58 @@ func (c *linuxContainer) newInitProcess(p *Process, cmd *exec.Cmd, parentPipe, c
+			nsPaths = append(nsPaths, ns.Path)
+			if ns.Type == configs.NEWPID {
+				doClone = true
+			}
```
You can't to enter into namespaces in a random order. For example, when you enter into a userns, you may lose permissions to enter in other namespaces.

avagin commented on container_linux.go:
```diff
@@ -169,21 +168,58 @@ func (c *linuxContainer) newInitProcess(p *Process, cmd *exec.Cmd, parentPipe, c
 
 func (c *linuxContainer) newSetnsProcess(p *Process, cmd *exec.Cmd, parentPipe, childPipe *os.File) *setnsProcess {
+	var nsPaths []string
+	for _, ns := range []string{"ipc", "uts", "net", "pid", "mnt"} {
```
Could you add "user" to this array?

dqminh replied:
Can I add that in the next patch ? I want to keep the behavior the same as now.

dqminh replied:
actually we probably can use `state.NamespacePaths` here instead of getting the proc path. We still need some code to sort that namespace paths so it enters by order, that code can be shared with the init process above.

avagin commented on container_linux.go:
```diff
@@ -169,21 +168,58 @@ func (c *linuxContainer) newInitProcess(p *Process, cmd *exec.Cmd, parentPipe, c
 func (c *linuxContainer) newSetnsProcess(p *Process, cmd *exec.Cmd, parentPipe, childPipe *os.File) *setnsProcess {
+	var nsPaths []string
+	for _, ns := range []string{"ipc", "uts", "net", "pid", "mnt"} {
+		path := fmt.Sprintf("/proc/%d/ns/%s", c.initProcess.pid(), ns)
```
The init process can died, but this code will not notice this.

dqminh replied:
right, i think in the old way we check for the existence of `/proc/pid/ns` to determine if the init is alive. I guess we can do the same here.

avagin replied:
In the old way we open("/proc/pid/ns") and use fstatat(fd, ns).
We call stat to determine whether a namespace is supported by kernel. We can do lstat("/proc/self/%s", ns) for this.

dqminh replied:
in `orderNamespacePaths`, we check for Lstat of the namespaces and fail if it doesnt exist.

avagin commented on process_linux.go:
```diff
@@ -226,7 +266,14 @@ func (p *initProcess) wait() (*os.ProcessState, error) {
+			joinPidns = true
+		}
+	}
+	if !joinPidns && p.cmd.SysProcAttr.Cloneflags&syscall.CLONE_NEWPID == 0 {
```
CLONE_NEWPID is set only when a process doesn't join another pidns, isn't it?

dqminh replied:
Yes. Actually i think this code is not needed, since it should be safe to kill the current container's processes because two containers dont share cgroups. I will do some test and remove this if it's the case. Thanks for noticing it !

dqminh replied:
done. The error I noticed was because two container in tests share the same cgroups. Changed the tests to use different cgroups for the containers.

avagin commented on nsenter/nsenter_test.go:
```diff
@@ -1,91 +0,0 @@
-	Pid int `json:"Pid"`
-}
-
-func TestNsenterAlivePid(t *testing.T) {
```
Why do you think these tests are useless?

dqminh replied:
Sorry i forgot to restore it when i changed the env. It was deleted because it has to be rewritten to test for invalid namespaces path, not invalid pids anymore since we only care about the namespace paths now.

I will restore it with new environment variables.

dqminh replied:
done. I restored nsenter_test with 2 tests for valid and invalid namespaces.

dqminh commented:
@avagin I pushed a new patch that unifies the ordering of namespaces in init and exec, and a few more changes. PTAL.

Also i notice that joining mount namespace for init process will not work right now because it's still trying to setupRootfs. I can write a new PR to fix that ( if we want a init process to join mnt namespace of another container ? )

dqminh committed:
validate namespace paths before saving it to list

- check if namespace is supported by the current kernel
- check that the pathname doesnt contain comma `,`. This is the character we
  used to join paths together and pass it over as a single env variable so we
  can split it out in C layer.

Signed-off-by: Daniel, Dao Quang Minh <dqminh89@gmail.com>

avagin commented on container_linux.go:
```diff
@@ -169,21 +168,60 @@ func (c *linuxContainer) newInitProcess(p *Process, cmd *exec.Cmd, parentPipe, c
+	if len(nsMaps) > 0 {
+		nsPaths, err := orderNamespacePaths(nsMaps)
+		if err != nil {
+			return nil, newSystemError(err)
```
orderNamepsapcePaths() returns SystemError

dqminh replied:
done

avagin commented on container_linux.go:
```diff
@@ -777,3 +815,37 @@ func (c *linuxContainer) currentState() (*State, error) {
+		configs.NEWNET,
+		configs.NEWPID,
+		// mnt namespace must be the last one. After switching to mnt
+		// namespace, the old /proc will be inaccessible.
```
I think USERNS must be the last one;). We will need to open a file descriptor for it before switching the mount namespace.

dqminh replied:
hmm yes i think so. WDYT about opening all file descriptors first before `setns` so it's consistent. I think doing that helps with ordering too (i.e., order doesnt matter). Something like:

```
fds = []int{}
for ns in namespaces { fds = append(fds, open(ns) }
for fd in fds { setns(fd, 0) }
```

avagin replied:
I think it's ok. But userns must be the last one, otherwise you will not have permissions to switch other namespaces.
https://github.com/torvalds/linux/blob/master/fs/namespace.c#L3235

dqminh replied:
added. Joining user namespace will only be applicable for execin process for now. There are some trouble with EPERM when setting up mnt/net when i was trying to let an init process join another init process in test. I didnt debug much further.

I figure that joining pid namespace is the most obvious usecase now. I can look into other usecase after (mnt/userns) if they make sense.

dqminh committed:
open all ns descriptors before setns

Doing this helps with ordering of NEWNS, since we dont have to worry about
set mount namespace too early leadings to old /proc becomes inaccessible.

Signed-off-by: Daniel, Dao Quang Minh <dqminh89@gmail.com>

dqminh committed:
add NEWUSER to supported namespaces to be joined

execin process can join user namespace of the init process.

Signed-off-by: Daniel, Dao Quang Minh <dqminh89@gmail.com>

dqminh committed:
join userns only for execin process

execin process can join user namespace of the parent container only when
the init process has joined user namespace explicitly.

Signed-off-by: Daniel, Dao Quang Minh <dqminh89@gmail.com>

avagin commented on container_linux.go:
```diff
@@ -777,3 +815,40 @@ func (c *linuxContainer) currentState() (*State, error) {
+		configs.NEWNS,
+	}
+	// For now, only join user namespace if this is an exec in process and the
+	// container supports user namespace
```
I would like to have explanation about this restriction.

avagin commented on nsenter/nsexec.c:
```diff
@@ -105,47 +110,64 @@ void nsexec()
+		nspaths = NULL;
+	}
+	for (i = 0; i < nsLen; i++) {
+		if (fds[i] != -1 && setns(fds[i], 0) != 0) {
```
fds[i] can't be negative here, cat it be?

avagin commented:
LGTM (except a few minor comments).

hqhq commented:
@dqminh Can you backport this PR to runc? Thanks.

dqminh commented:
@hqhq Thanks for reminding me !. Things are a bit hectic on my side this week, but i should be able get it done by the weekend (or sooner if possible)

LK4D4 commented:
ported to runc
